### PR TITLE
[fix] 'from' in python code generated by inductor, resulting in SyntaxError

### LINKED ISSUE #121621: torch.Tensor.random_ causes invalid syntax in InternalTorchDynamoError
### 🐛 Describe the bug

The op signature from https://pytorch.org/docs/stable/generated/torch.Tensor.random_.html#torch.Tensor.random_ 
is that first argument is from and second to, looks like this is causing trouble in tracing

### Error logs

```
Traceback (most recent call last):
  File "/tmp/add.py", line 13, in <module>
    res = random_op(tensor, params)
  File "/tmp/lib/python3.10/site-packages/torch/_dynamo/eval_frame.py", line 489, in _fn
    return fn(*args, **kwargs)
  File "/tmp/lib/python3.10/site-packages/torch/_dynamo/eval_frame.py", line 655, in catch_errors
    return callback(frame, cache_entry, hooks, frame_state)
  File "/tmp/lib/python3.10/site-packages/torch/_dynamo/convert_frame.py", line 727, in _convert_frame
    result = inner_convert(frame, cache_entry, hooks, frame_state)
  File "/tmp/lib/python3.10/site-packages/torch/_dynamo/convert_frame.py", line 383, in _convert_frame_assert
    compiled_product = _compile(
  File "/tmp/lib/python3.10/site-packages/torch/_dynamo/convert_frame.py", line 665, in _compile
    raise InternalTorchDynamoError(str(e)).with_traceback(
  File "/tmp/lib/python3.10/site-packages/torch/_dynamo/convert_frame.py", line 646, in _compile
    guarded_code = compile_inner(code, one_graph, hooks, transform)
  File "/tmp/lib/python3.10/site-packages/torch/_dynamo/utils.py", line 244, in time_wrapper
    r = func(*args, **kwargs)
  File "/tmp/lib/python3.10/site-packages/torch/_dynamo/convert_frame.py", line 562, in compile_inner
    out_code = transform_code_object(code, transform)
  File "/tmp/lib/python3.10/site-packages/torch/_dynamo/bytecode_transformation.py", line 1033, in transform_code_object
    transformations(instructions, code_options)
  File "/tmp/lib/python3.10/site-packages/torch/_dynamo/convert_frame.py", line 151, in _fn
    return fn(*args, **kwargs)
  File "/tmp/lib/python3.10/site-packages/torch/_dynamo/convert_frame.py", line 527, in transform
    tracer.run()
  File "/tmp/lib/python3.10/site-packages/torch/_dynamo/symbolic_convert.py", line 2128, in run
    super().run()
  File "/tmp/lib/python3.10/site-packages/torch/_dynamo/symbolic_convert.py", line 818, in run
    and self.step()
  File "/tmp/lib/python3.10/site-packages/torch/_dynamo/symbolic_convert.py", line 781, in step
    getattr(self, inst.opname)(inst)
  File "/tmp/lib/python3.10/site-packages/torch/_dynamo/symbolic_convert.py", line 2243, in RETURN_VALUE
    self.output.compile_subgraph(
  File "/tmp/lib/python3.10/site-packages/torch/_dynamo/output_graph.py", line 919, in compile_subgraph
    self.compile_and_call_fx_graph(tx, list(reversed(stack_values)), root)
  File "/usr/lib/python3.10/contextlib.py", line 79, in inner
    return func(*args, **kwds)
  File "/tmp/lib/python3.10/site-packages/torch/_dynamo/output_graph.py", line 1063, in compile_and_call_fx_graph
    gm = fx.GraphModule(root, self.graph)
  File "/tmp/lib/python3.10/site-packages/torch/fx/graph_module.py", line 428, in __init__
    self.graph = graph
  File "/tmp/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1750, in __setattr__
    super().__setattr__(name, value)
  File "/tmp/lib/python3.10/site-packages/torch/fx/graph_module.py", line 472, in graph
    self.recompile()
  File "/tmp/lib/python3.10/site-packages/torch/fx/graph_module.py", line 724, in recompile
    cls.forward = _forward_from_src(self._code, python_code.globals, co_fields)
  File "/tmp/lib/python3.10/site-packages/torch/_dynamo/convert_frame.py", line 122, in fx_forward_from_src_skip_result
    result: types.FunctionType = original_forward_from_src(*args, **kwargs)
  File "/tmp/lib/python3.10/site-packages/torch/fx/graph_module.py", line 88, in _forward_from_src
    return _method_from_src(
  File "/tmp/lib/python3.10/site-packages/torch/fx/graph_module.py", line 98, in _method_from_src
    _exec_with_source(src, globals_copy, co_fields)
  File "/tmp/lib/python3.10/site-packages/torch/fx/graph_module.py", line 84, in _exec_with_source
    exec(compile(src, key, "exec"), globals)
torch._dynamo.exc.InternalTorchDynamoError: invalid syntax (<eval_with_key>.0, line 6)
```

### Minified repro

```
import torch


def random_op(tensor, params):
    res = tensor.random_(**params)
    return res


if __name__ == "__main__":
    random_op = torch.compile(random_op)
    params = {"from": -10, "to": 10}
    tensor = torch.randn([2, 3])
    res = random_op(tensor, params)
    print(res)
```

### Versions

[pip3] numpy==1.26.4
[pip3] torch==2.2.1
[pip3] torchaudio==2.2.0
[pip3] torchdata==0.7.1
[pip3] torchmetrics==1.2.1
[pip3] torchtext==0.17.0
[pip3] torchvision==0.17.0

cc @ezyang @msaroufim @bdhirsh @anijain2305 @zou3519 @chauhang @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng

### PR DESCRIPTION
**Pull request by LamForest**
`from` is  a python keyword.  Python code containing `aten.random.from` will result in syntax errors. This PR handle this special case in the codegen of inductor. 

Also add fallback lowering for aten.random

Fixes #121621 

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

**Comment by LamForest** [comment]
I replace `aten.random_.from` with `getattr(aten.random_, "from")`. This is a workaround. Please give me a hint if anyone has a better solution.

IMO, we should rename `random_.from` to `random_.from_to` to avoid simliar issues. This issue can appear in all python codegen codes.

**Review by jansel — CHANGES_REQUESTED** [review]
@LamForest this fix looks good to me, but can you add a test for it?

**Comment by LamForest** [comment]
@jansel Sure, please enlighten me where to put my test

**Comment by jansel** [comment]
You can add it here: https://github.com/pytorch/pytorch/blob/main/test/inductor/test_torchinductor.py

File: torch/_inductor/codegen/wrapper.py [hunk 1]
```diff
@@ -634,6 +634,13 @@ def generate_extern_kernel_alloc(self, extern_kernel, args):
             # view operation fallbacks cause issues since inductor
             # doesn't know the memory is still needed and might reuse it.
             ending = f".clone(){ending}"
+
+        # Note: 'from' is python keyword, it is not allowed to appear in python code
+        #       Although, only aten.random_ has an 'from' overload for now, we handle all possible cases
+        if kernel_name.endswith(".from"):
```
Comment (Skylion007) [inline torch/_inductor/codegen/wrapper.py:640]: Can we generalize this handling by using https://docs.python.org/3/library/keyword.html ?

**Review by jansel — APPROVED** [review]

File: torch/_inductor/codegen/wrapper.py [hunk 2]
```diff
@@ -203,9 +204,9 @@ def __str__(self):
 class MemoryPlanningState:
     def __init__(self):
         super().__init__()
-        self.reuse_pool: Dict[
-            ReuseKey, List[FreeIfNotReusedLine]
-        ] = collections.defaultdict(list)
+        self.reuse_pool: Dict[ReuseKey, List[FreeIfNotReusedLine]] = (
+            collections.defaultdict(list)
+        )
```
Comment (jansel) [inline torch/_inductor/codegen/wrapper.py:209]: This change is causing lints to fail:
https://hud.pytorch.org/pr/pytorch/pytorch/122632#23375048104

**Review by jansel — CHANGES_REQUESTED** [review]
Looks like the new test is failing on CI

File: torch/_inductor/codegen/wrapper.py (hunk 2 above)
Comment (LamForest) [reply torch/_inductor/codegen/wrapper.py:209]: Emm, this is changed by runnning `lintrunner -f` locally. Let me revert this change

**Comment by LamForest** [comment]
Need some time to look at those CI failures..

File: torch/_inductor/codegen/wrapper.py (hunk 2 above)
Comment (jansel) [reply torch/_inductor/codegen/wrapper.py:209]: Maybe wrong versions.  You can do `make setup_lint`.

**Pull request closed without merging**